File: Striver/Recursion/q12.java
package Striver.Recursion;

public class q12 {
    // Print a single subsequence which sums to k
    public static boolean printSubsequence(int[] arr, int k, int i, String ans) {
        if (i == arr.length) {
            if (k == 0) {
                System.out.println(ans);
                return true;
            }
            return false;
        }
        if(printSubsequence(arr, k - arr[i], i + 1, ans + arr[i] + " ")==true) return true;
        if(printSubsequence(arr, k, i + 1, ans)==true) return true;
        return false;
    }

    public static void main(String[] args) {
        int[] arr = { 1, 2, 3, 4, 5, 6, 7, 8, 9 };
        int k = 10;
        printSubsequence(arr, k, 0, "");
    }
}

File: Striver/Recursion/q12.py
# Print a single subsequence which sums to k

def printSubsequences(arr, k, output, index, sum):
    if index == len(arr):
        if sum == k:
            print(output)
            return True
        return False
    if printSubsequences(arr, k, output, index + 1, sum) == True:
        return True
    output.append(arr[index])
    if printSubsequences(arr, k, output, index + 1, sum + arr[index]) == True:
        return True
    output.pop()
    return False

arr = [1, 2, 3, 4, 5]
printSubsequences(arr, 5, [], 0, 0)